### PR TITLE
fix!: use type module (revert #1411)

### PR DESCRIPTION
**Pull request by bluwy**
Reverts #1411 (and also #1439)

Ref: https://github.com/vitest-dev/vitest/pull/1411#issuecomment-1151632260

It's not needed to remove type module to fix the issue stated in the PR

### CODE REVIEW

**Comment by sheremet-va** [comment]
Will it revert errors for users? If so, what should we recommend people?

**Comment by bluwy** [comment]
Following https://github.com/vitest-dev/vitest/pull/1411#issuecomment-1147530146, if they're using the new `nodenext` or `node16` options, they must name the test files with `.mts` if `"type": "module"` is not specified.

**Comment by sheremet-va** [comment]
> Yes it will revert the errors. Following [#1411 (comment)](https://github.com/vitest-dev/vitest/pull/1411#issuecomment-1147530146), if they're using the new `nodenext` or `node16` options, they must name the test files with `.mts` if `"type": "module"` is not specified.

In that case, should we also expand our `default` our defaults?

https://github.com/vitest-dev/vitest/blob/fef00a7328907b7663fa37e06cffa1781b84cd88/packages/vitest/src/defaults.ts#L6

https://github.com/vitest-dev/vitest/blob/fef00a7328907b7663fa37e06cffa1781b84cd88/packages/vitest/src/defaults.ts#L15-L16

**Comment by bluwy** [comment]
I think that makes sense 👍

**Comment by antfu** [comment]
I would hold this for a while longer. Indeed the usage in https://github.com/vitest-dev/vitest/pull/1411 might better use `.mts` for the tests. But as `.mts` is not wildly supported yet, plus there is not too much harm for us to use `.mjs`, I would more lean to make it easy for users until we really encounter some blockers for the current workaround.

**Comment by aleclarson** [comment]
> Following [#1411 (comment)](https://github.com/vitest-dev/vitest/pull/1411#issuecomment-1147530146), if they're using the new `nodenext` or `node16` options, they must name the test files with `.mts` if `"type": "module"` is not specified.

The "better solution" I mentioned in that thread goes like this:
```json
".": {
  "require": {
    "types": "./dist/index.d.cts",
    "default": "./dist/index.cjs"
  },
  "types": "./dist/index.d.ts",
  "import": "./dist/index.js"
},
```

That would be how Vitest's `exports` should look. Then users aren't required to change their setup at all.

Both `dist/index.d.cts` and `dist/index.cjs` can just re-export from the other dist files.

**Comment by sheremet-va** [comment]
> The "better solution" I mentioned in that thread goes like this:

First of all, we already saw in that PR that it doesn't work, does it? TypeScript has a big preventing that.

Second of all, I don't see how this is a "better solution"? We will essentially double package size just to please few people with TypeScript issues that can be resolved on their end? Those files won't even be run in any situation.

**Comment by aleclarson** [comment]
> First of all, we already saw in that PR that it doesn't work, does it? TypeScript has a big preventing that.

The two differences from your idea are that (1) `require.types` is defined and (2) it uses a `.d.cts` extension. Both are required for TypeScript to believe the package is not ESM only.



> We will essentially double package size

No, the new `.cts` and `.cjs` files are basic re-exports of the ESM dist files.

```ts
// dist/index.d.cts
export * from './index.mjs'

// dist/index.cjs
export * from './index.mjs'
```

**Comment by aleclarson** [comment]
> But this wouldn't work for the same reason it doesn't work right now, no? You cannot use static import/export inside cjs.

Those modules would never be loaded by Node.js runtime, so no errors would occur.

Correction: The `dist/index.cjs` module should actually be empty. The `dist/index.d.cts` allows import/export syntax, and I've tested it myself, so I know it works. :)

It's not a "hack", because the TypeScript team themselves said it's a valid solution.

**Comment by sheremet-va** [comment]
> Those modules would never be loaded by Node.js runtime, so no errors would occur.

No, I mean by TypeScript. Why would `importing` ESM from `cts` would work inside a node_module, but not in your own project?

Current problem:
- importing Vitest from `commonjs` gives error from TypeScript typechecker, because it will compile `import` to `require` and `requiring` Vitest is not allowed
- but `importing` inside Vitest's `cts` magically works? I don't buy it tbh

**Comment by sheremet-va** [comment]
Ok, I've looked through the threads, and now I see that @aleclarson's idea is valid. We can include it in this PR actually, and safely merge it. @antfu, what do you think?

**Comment by antfu** [comment]
I am not very familiar with this topic, but sure, if it helps and doesn't break anything else :)

File: packages/vitest/rollup.config.js
Comment (sheremet-va) [inline packages/vitest/rollup.config.js:81]: I think we can just put this files in root?

**Review by sheremet-va — CHANGES_REQUESTED** [review]

File: packages/vitest/rollup.config.js
Comment (bluwy) [reply packages/vitest/rollup.config.js:81]: That's possible too if you don't mind the two extra files in root. I can change that.

File: packages/vitest/package.json
Comment (bluwy) [inline packages/vitest/package.json:32]: Note: I had to alter this a bit as the eslint config sorts the `types` key above `require`.

Comment (bluwy) [reply packages/vitest/package.json:32]: The rule comes from https://github.com/antfu/eslint-config/blob/8ddaee3bf8a942b4bc73cd009daf6b67f155c1ad/packages/basic/index.js#L125-L130

File: packages/vitest/rollup.config.js
Comment (bluwy) [reply packages/vitest/rollup.config.js]: Updated 👍 I kept the `defineConfig` since it provides types. But let me know if you want to minimize the changes here.

Comment (sheremet-va) [reply packages/vitest/rollup.config.js]: I'm fine with `defineConfig`, good job 👍🏻

**Review by sheremet-va — APPROVED** [review]

**Comment by jroru** [comment]
Is there anything blocking this from being merged?

**Comment by sheremet-va** [comment]
> Is there anything blocking this from being merged?

This needs updating, we changed quite a few APIs since then, and if merged, Vitest won't work.

**Comment by jroru** [comment]
For a package with:

* `package.json` `type` set to `module`
* `tsconfig.json` `compilerOptions.moduleResolution` set to `Node16`
* `vitest@0.22.1`
* `vite@3.0.9`
* `typescript@4.7.4`

Running `tsc` outputs:

```
../../node_modules/vitest/dist/config.d.ts:2:27 - error TS1471: Module 'vite' cannot be imported using this construct. The specifier only resolves to an ES module, which cannot be imported synchronously. Use dynamic import instead.

2 export { ConfigEnv } from 'vite'
```

Is there a known workaround?

**Comment by bluwy** [comment]
I've updated the PR. Tried to update some of the newer places, and the tests seems to be passing, except ubuntu node14 (not sure if it's a fluke).

**Comment by StevenGBrown** [comment]
> For a package with:
> 
> * `package.json` `type` set to `module`
> * `tsconfig.json` `compilerOptions.moduleResolution` set to `Node16`
> * `vitest@0.22.1`
> * `vite@3.0.9`
> * `typescript@4.7.4`
> 
> Running `tsc` outputs:
> 
> ```
> ../../node_modules/vitest/dist/config.d.ts:2:27 - error TS1471: Module 'vite' cannot be imported using this construct. The specifier only resolves to an ES module, which cannot be imported synchronously. Use dynamic import instead.
> 
> 2 export { ConfigEnv } from 'vite'
> ```
> 
> Is there a known workaround?

@jroru My workaround until this PR is merged has been to use `patch-package` to add `"type": "module"` to the vitest package.json.

**Pull request merged**
